File: server/src/utils/passwordValidator.ts
import db from '../db/database';

export interface PasswordRules {
  minLength: number;
  requireDigit: boolean;
  requireSpecial: boolean;
}

export function getPasswordRules(): PasswordRules {
  const rows = db.prepare(
    "SELECT key, value FROM settings WHERE key LIKE 'pw_%'"
  ).all() as { key: string; value: string }[];
  const map = Object.fromEntries(rows.map(r => [r.key, r.value]));
  return {
    minLength:      Number(map['pw_min_length'] ?? 6),
    requireDigit:   map['pw_require_digit']   === '1',
    requireSpecial: map['pw_require_special'] === '1',
  };
}

export function validatePassword(password: string, rules: PasswordRules): string | null {
  if (password.length < rules.minLength) {
    return `Passwort muss mindestens ${rules.minLength} Zeichen lang sein`;
  }
  if (rules.requireDigit && !/\d/.test(password)) {
    return 'Passwort muss mindestens eine Ziffer enthalten';
  }
  if (rules.requireSpecial && !/[^a-zA-Z0-9]/.test(password)) {
    return 'Passwort muss mindestens ein Sonderzeichen enthalten';
  }
  return null;
}

File: server/utils/password_validator.py
"""Password rules helper — equivalent of passwordValidator.ts."""
import re
from dataclasses import dataclass
from db.database import get_db


@dataclass
class PasswordRules:
    min_length: int
    require_digit: bool
    require_special: bool


def get_password_rules() -> PasswordRules:
    """Read current password rules from the settings table."""
    conn = get_db()
    rows = conn.execute(
        "SELECT key, value FROM settings WHERE key LIKE 'pw_%'"
    ).fetchall()
    mapping = {r["key"]: r["value"] for r in rows}
    return PasswordRules(
        min_length=int(mapping.get("pw_min_length", 6)),
        require_digit=mapping.get("pw_require_digit", "0") == "1",
        require_special=mapping.get("pw_require_special", "0") == "1",
    )


def validate_password(password: str, rules: PasswordRules) -> str | None:
    """Return an error message string, or None if the password is valid."""
    if len(password) < rules.min_length:
        return f"Passwort muss mindestens {rules.min_length} Zeichen lang sein"
    if rules.require_digit and not re.search(r"\d", password):
        return "Passwort muss mindestens eine Ziffer enthalten"
    if rules.require_special and not re.search(r"[^a-zA-Z0-9]", password):
        return "Passwort muss mindestens ein Sonderzeichen enthalten"
    return None
